messaging/src/main/java/io/lantern/messaging/CryptoWorker.kt

```diff
@@ -0,0 +1,837 @@
+package io.lantern.messaging
+
+import io.lantern.db.ChangeSet
+import io.lantern.db.Subscriber
+import io.lantern.db.Transaction
+import io.lantern.messaging.tassis.Callback
+import io.lantern.messaging.tassis.InboundMessage
+import io.lantern.messaging.tassis.Messages
+import io.lantern.messaging.tassis.Padding
+import io.lantern.messaging.tassis.byteString
+import io.lantern.messaging.time.minutesToMillis
+import okhttp3.Call
+import okhttp3.MultipartBody
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okhttp3.RequestBody.Companion.asRequestBody
+import okhttp3.Response
+import org.signal.libsignal.metadata.SealedSessionCipher
+import org.whispersystems.libsignal.DeviceId
+import org.whispersystems.libsignal.SessionBuilder
+import org.whispersystems.libsignal.SignalProtocolAddress
+import org.whispersystems.libsignal.ecc.ECPublicKey
+import org.whispersystems.libsignal.state.PreKeyBundle
+import org.whispersystems.libsignal.state.PreKeyRecord
+import org.whispersystems.libsignal.state.SignedPreKeyRecord
+import org.whispersystems.signalservice.api.crypto.AttachmentCipherOutputStream
+import org.whispersystems.signalservice.internal.util.Util
+import java.io.File
+import java.io.FileOutputStream
+import java.io.IOException
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+
+/**
+ * CryptoWorker handles all sending and receiving of messages.
+ *
+ * Outbound messages are modeled as a work queue consisting of OutboundMessages. All encryption of
+ * outbound messages happens on a single worker thread, with another worker thread dedicated to
+ * encrypting attachments.
+ */
+internal class CryptoWorker(
+    messaging: Messaging,
+    retryDelayMillis: Long,
+    private val stopSendRetryAfterMillis: Long
+) :
+    Worker(messaging, "crypto", retryDelayMillis = retryDelayMillis) {
+    private val db = messaging.db
+    private val store = messaging.store
+    private val httpClient = OkHttpClient() // TODO: configure support for proxying and stuff
+    private val cipher = SealedSessionCipher(store, store.deviceId)
+    private val uploadAuthorizations = ArrayDeque<Messages.UploadAuthorization>()
+    private val encryptAttachmentsExecutor = Executors.newSingleThreadExecutor()
+
+    init {
+        // find out about all disappearing messages (including previously saved ones) and schedule
+        // them for deletion
+        db.subscribe(
+            object : Subscriber<String>(
+                "disappearingMessagesSubscriber",
+                Schema.PATH_DISAPPEARING_MESSAGES.path('%')
+            ) {
+                override fun onChanges(changes: ChangeSet<String>) {
+                    changes.updates.forEach { (path, msgPath) ->
+                        val scheduledTimeMillis = path.split("/")[2].toLong()
+                        val delayMillis = scheduledTimeMillis - now
+                        executor.schedule(
```

Comment:
No, because it's scheduled based on what we're reading from the database. If the app is closed, the next time we open the app we'll get back into this code and schedule it again.